### PR TITLE
chain: Add result to genesis

### PR DESCRIPTION
**Pull request by perimeko**

### CODE REVIEW

File: chain/x/oracle/keeper/result_test.go [hunk 1]
```diff
@@ -69,3 +78,60 @@ package keeper
 // 	err = keeper.AddResult(ctx, 1, 1, []byte("calldata"), []byte("result"))
 // 	require.NotNil(t, err)
 // }
+
+func TestGetAllResults(t *testing.T) {
+	_, ctx, k := createTestInput()
+
+	reqPacket1 := oracle.OracleRequestPacketData{
```
Comment (sorawit) [inline chain/x/oracle/keeper/result_test.go]: Let's use New factory function (`NewOracleRequestPacketData`)

File: chain/x/oracle/keeper/result_test.go [hunk 2]
```diff
@@ -69,3 +78,60 @@ package keeper
...
+		Result:        BasicCalldata,
+	}
+
+	_, err = k.AddResult(ctx, types.RequestID(4), reqPacket4, resPacket4)
+	require.NoError(t, err)
+
+	results := k.GetAllResults(ctx)
+
+	require.Equal(t, 4, len(results))
+	require.NotEmpty(t, results[0])
```
Comment (sorawit) [inline chain/x/oracle/keeper/result_test.go]: Let's check for the exact value.

File: chain/x/oracle/keeper/result.go [hunk 3]
```diff
@@ -50,3 +50,24 @@ func (k Keeper) AddResult(
 	store.Set(types.ResultStoreKey(id), resultHash)
 	return resultHash, nil
 }
+
+// GetAllResults return the list of all results in the store, or nil if there is none
```
Comment (sorawit) [inline chain/x/oracle/keeper/result.go]: GetAllResults returns the list of all results in the store. Nil will be added for skipped results.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/genesis.go [hunk 4]
```diff
@@ -73,6 +77,10 @@ func InitGenesis(ctx sdk.Context, k Keeper, data GenesisState) []abci.ValidatorU
 		k.SetValidatorReportInfo(ctx, info.Validator, info)
 	}
 
+	for i, result := range data.Results {
+		k.SetResult(ctx, types.RequestID(i+1), result)
```
Comment (sorawit) [inline chain/x/oracle/genesis.go]: Let's skip if result is nil

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/genesis.go (hunk 4 above)
Comment (perimeko) [reply chain/x/oracle/genesis.go]: fixed

File: chain/x/oracle/keeper/result_test.go (hunk 1 above)
Comment (perimeko) [reply chain/x/oracle/keeper/result_test.go]: fixed

File: chain/x/oracle/keeper/result.go [hunk 5]
```diff
@@ -50,3 +50,30 @@ func (k Keeper) AddResult(
 	store.Set(types.ResultStoreKey(id), resultHash)
 	return resultHash, nil
 }
+
+// SetResult set result to the store.
+func (k Keeper) SetResult(ctx sdk.Context, regID types.RequestID, result []byte) {
```
Comment (sorawit) [inline chain/x/oracle/keeper/result.go]: Move it to above `AddResult`. Also please name variable `reqID`

File: chain/x/oracle/keeper/result.go [hunk 6]
```diff
@@ -50,3 +50,30 @@ func (k Keeper) AddResult(
 	store.Set(types.ResultStoreKey(id), resultHash)
```
Comment (sorawit) [inline chain/x/oracle/keeper/result.go]: Now that we have `SetResult`. I think this should call into `SetResult`.

File: chain/x/oracle/keeper/result_test.go [hunk 7]
```diff
@@ -69,3 +76,47 @@ package keeper
...
+	require.Equal(t, resultHashReqID4, results[3])
+}
+
+func TestSetResult(t *testing.T) {
+	_, ctx, k := createTestInput()
+
+	reqPacket := types.NewOracleRequestPacketData("alice", 1, BasicCalldata, 1, 1)
+	resPacket := types.NewOracleResponsePacketData("alice", 1, 1, 1589535020, 1589535022, 1, BasicCalldata)
+
+	resultHash, err := k.AddResult(ctx, types.RequestID(1), reqPacket, resPacket)
```
Comment (sorawit) [inline chain/x/oracle/keeper/result_test.go:113]: Let's also test `GetResult` here

File: chain/x/oracle/keeper/result_test.go [hunk 8]
```diff
@@ -69,3 +76,47 @@ package keeper
...
+
+	resultHashReqID4, err := k.AddResult(ctx, types.RequestID(4), reqPacket4, resPacket4)
+	require.NoError(t, err)
+
+	results := k.GetAllResults(ctx)
+
+	require.Equal(t, 4, len(results))
+	require.Equal(t, resultHashReqID1, results[0])
+
+	// result of reqID 2 and 3 should be empty byte array
```
Comment (sorawit) [inline chain/x/oracle/keeper/result_test.go]: It should be nil, not empty byte array

File: chain/x/oracle/keeper/result.go [hunk 9]
```diff
@@ -50,3 +50,30 @@ func (k Keeper) AddResult(
...
+
+// GetAllResults returns the list of all results in the store. Nil will be added for skipped results.
+func (k Keeper) GetAllResults(ctx sdk.Context) (results [][]byte) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.ResultStoreKeyPrefix)
+	var previousReqID types.RequestID
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		currentReqID := types.RequestID(sdk.BigEndianToUint64(iterator.Key()[1:]))
+		diffReqIDCount := currentReqID - previousReqID
```
Comment (sorawit) [inline chain/x/oracle/keeper/result.go]: let's use `:= int(...)` here then we don't need to do casting in the for statement.

File: chain/x/oracle/keeper/result.go [hunk 10]
```diff
@@ -50,3 +50,30 @@ func (k Keeper) AddResult(
...
+
+		// Insert nil for the request that result is empty
```
Comment (sorawit) [inline chain/x/oracle/keeper/result.go]: Insert nil for each request without result.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/result.go [hunk 11]
```diff
@@ -50,3 +50,30 @@ func (k Keeper) AddResult(
 	store.Set(types.ResultStoreKey(id), resultHash)
 	return resultHash, nil
 }
+
+// SetResult set result to the store.
+func (k Keeper) SetResult(ctx sdk.Context, reqID types.RequestID, result []byte) {
```
Comment (sorawit) [inline chain/x/oracle/keeper/result.go]: Let's move this function to above `GetResult` for consistency.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/result.go (hunk 5 above)
Comment (perimeko) [reply chain/x/oracle/keeper/result.go]: fixed

File: chain/x/oracle/keeper/result.go (hunk 6 above)
Comment (perimeko) [reply chain/x/oracle/keeper/result.go]: fixed, used set result function

File: chain/x/oracle/keeper/result_test.go (hunk 7 above)
Comment (perimeko) [reply chain/x/oracle/keeper/result_test.go:113]: fixed

File: chain/x/oracle/keeper/result.go (hunk 9 above)
Comment (perimeko) [reply chain/x/oracle/keeper/result.go]: fixed

**Review by sorawit — APPROVED** [review]

File: chain/x/oracle/keeper/result.go [hunk 12]
```diff
@@ -13,6 +13,12 @@ func (k Keeper) HasResult(ctx sdk.Context, id types.RequestID) bool {
 	return ctx.KVStore(k.storeKey).Has(types.ResultStoreKey(id))
 }
 
+// SetResult set result to the store.
```
Comment (sorawit) [inline chain/x/oracle/keeper/result.go]: sets

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**